File: src/types/framework.ts
// ─── Framework domain types ───────────────────────────────────────────────────

export type FrameworkSlug =
  | 'pomodoro'
  | 'time_blocking'
  | 'deep_work'
  | 'gtd'
  | 'daily_6'
  | 'eat_the_frog'
  | 'custom'

export type FrameworkState =
  | 'idle'
  | 'active'
  | 'paused'
  | 'completed'
  | 'interrupted'

export type PomodoroPhase = 'work' | 'short_break' | 'long_break'

export interface PomodoroState {
  phase: PomodoroPhase
  /** Elapsed time in milliseconds */
  elapsed: number
  /** Total duration of current phase in milliseconds */
  duration: number
  /** Current interval index (1-based) */
  interval: number
  /** Total intervals before long break (typically 4) */
  totalIntervals: number
}

export interface FrameworkStep {
  id: string
  title: string
  description?: string
  durationMinutes?: number
  isOptional: boolean
  actionType?: 'review' | 'plan' | 'work' | 'reflect' | 'break' | 'capture'
}

export interface FrameworkDefinition {
  slug: FrameworkSlug
  name: string
  description: string
  icon: string
  color: string
  steps: FrameworkStep[]
  defaultDurationMinutes: number
  supportsPomodoro: boolean
}

export interface Framework {
  id: string
  userId: string
  definitionSlug: FrameworkSlug
  name: string
  description?: string
  isCustom: boolean
  config?: CustomFrameworkConfig
  /** ISO-8601 dates for historical tracking */
  lastUsedAt?: string
  useCount: number
  createdAt: string
  updatedAt: string
}

export interface CustomFrameworkConfig {
  steps: FrameworkStep[]
  defaultDurationMinutes: number
  pomodoroWorkMinutes?: number
  pomodoroShortBreakMinutes?: number
  pomodoroLongBreakMinutes?: number
  intervalsBeforeLongBreak?: number
  autoStartBreaks?: boolean
  autoStartWork?: boolean
  notifyOnStepChange?: boolean
}

export interface FrameworkSession {
  id: string
  userId: string
  frameworkId: string
  frameworkSlug: FrameworkSlug
  state: FrameworkState
  currentStepId?: string
  pomodoroState?: PomodoroState
  linkedTaskIds: string[]
  /** ISO-8601 datetime */
  startedAt: string
  /** ISO-8601 datetime */
  pausedAt?: string
  /** ISO-8601 datetime */
  completedAt?: string
  totalPausedMs: number
  notes?: string
  completedIntervals: number
  createdAt: string
  updatedAt: string
}

File: src-tauri/src/models/framework.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Framework type / slug
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FrameworkType {
    Builtin,
    Custom,
}

/// Well-known framework slugs. `Custom(String)` handles user-defined ones.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FrameworkSlug {
    /// "Eat the Frog" daily 6-task prioritization
    Daily6,
    /// Pomodoro technique
    Pomodoro,
    /// Getting Things Done
    Gtd,
    /// Time-blocking day planner
    TimeBlocking,
    /// User-defined; inner string is the slug
    #[serde(untagged)]
    Custom(String),
}

impl FrameworkSlug {
    pub fn as_str(&self) -> &str {
        match self {
            FrameworkSlug::Daily6 => "daily6",
            FrameworkSlug::Pomodoro => "pomodoro",
            FrameworkSlug::Gtd => "gtd",
            FrameworkSlug::TimeBlocking => "time_blocking",
            FrameworkSlug::Custom(s) => s.as_str(),
        }
    }
}

// ---------------------------------------------------------------------------
// Framework definition (serialisable step graph)
// ---------------------------------------------------------------------------

/// A single step within a framework (state-machine node).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FrameworkStep {
    pub id: String,
    pub label: String,
    pub description: Option<String>,
    /// Duration in seconds (None = unbounded / user-driven)
    pub duration_secs: Option<u64>,
    /// IDs of steps this step can transition to
    pub next_steps: Vec<String>,
    /// Arbitrary metadata consumed by the engine (e.g. {"pomodoro_type": "work"})
    pub meta: Option<JsonValue>,
}

/// The full declarative definition of a framework.
/// Stored as JSON in the app database so custom frameworks can be loaded without code changes.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FrameworkDefinition {
    pub slug: FrameworkSlug,
    pub name: String,
    pub description: String,
    pub version: String,
    pub initial_step: String,
    pub steps: Vec<FrameworkStep>,
    /// Configurable parameters exposed to the user
    pub config_schema: Option<JsonValue>,
}

// ---------------------------------------------------------------------------
// Runtime session
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FrameworkSession {
    pub id: Uuid,
    pub framework_id: Uuid,
    pub framework_slug: FrameworkSlug,
    /// Task IDs the user chose to work on during this session
    pub task_ids: Vec<Uuid>,
    pub state: FrameworkState,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    /// User-provided config overrides (e.g. {"work_mins": 50})
    pub config: Option<JsonValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FrameworkState {
    pub current_step: String,
    pub step_started_at: DateTime<Utc>,
    /// Seconds elapsed in the current step
    pub elapsed_secs: u64,
    /// Pomodoro-specific state (None for non-Pomodoro frameworks)
    pub pomodoro: Option<PomodoroState>,
    /// Arbitrary key-value store for framework engine extensions
    pub context: JsonValue,
}

// ---------------------------------------------------------------------------
// Pomodoro specifics
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PomodoroInterval {
    Work,
    ShortBreak,
    LongBreak,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PomodoroState {
    pub interval: PomodoroInterval,
    /// Number of completed work intervals so far in this session
    pub completed_intervals: u32,
    /// After how many work intervals a long break is triggered (default 4)
    pub long_break_after: u32,
    /// Duration overrides in seconds
    pub work_secs: u64,
    pub short_break_secs: u64,
    pub long_break_secs: u64,
}

impl Default for PomodoroState {
    fn default() -> Self {
        PomodoroState {
            interval: PomodoroInterval::Work,
            completed_intervals: 0,
            long_break_after: 4,
            work_secs: 25 * 60,
            short_break_secs: 5 * 60,
            long_break_secs: 15 * 60,
        }
    }
}

// ---------------------------------------------------------------------------
// Framework (the metadata record stored in the DB / config)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Framework {
    pub id: Uuid,
    pub framework_type: FrameworkType,
    pub definition: FrameworkDefinition,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Framework {
    pub fn new(definition: FrameworkDefinition, framework_type: FrameworkType) -> Self {
        let now = Utc::now();
        Framework {
            id: Uuid::new_v4(),
            framework_type,
            definition,
            is_active: false,
            created_at: now,
            updated_at: now,
        }
    }
}
